### PR TITLE
Adding fromXContent to Suggestion.Entry and subclasses

### PR DESCRIPTION
**Pull request by cbuescher**
This adds parsing from xContent to Suggestion.Entry and its subclasses for Terms-, Phrase- 
and CompletionSuggestion.Entry.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/suggest/Suggest.java [hunk 1]
```diff
@@ -371,37 +372,38 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-
-                static final String TEXT = "text";
-                static final String OFFSET = "offset";
-                static final String LENGTH = "length";
-                static final String OPTIONS = "options";
-
-            }
+            protected static final String TEXT = "text";
+            protected static final String OFFSET = "offset";
+            protected static final String LENGTH = "length";
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: I think some of them could be private

File: core/src/main/java/org/elasticsearch/search/suggest/Suggest.java [hunk 2]
```diff
@@ -551,6 +553,24 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 return builder;
             }
 
+            private static ObjectParser<Entry<Option>, Void> PARSER = new ObjectParser<>("SuggestionEntryParser",
+                    true, Entry::new);
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: Can you remove the line break?

File: core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestion.java [hunk 3]
```diff
@@ -100,6 +103,18 @@ public void addOption(Suggestion.Entry.Option option) {
             }
         }
 
+        private static ObjectParser<Entry, Void> PARSER = new ObjectParser<>("PhraseSuggestionEntryParser", true,
+                Entry::new);
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/suggest/phrase/PhraseSuggestion.java]: It should fit on 1 line

File: core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestion.java [hunk 4]
```diff
@@ -154,6 +155,18 @@ protected Option newOption() {
             return new Option();
         }
 
+        private static ObjectParser<Entry, Void> PARSER = new ObjectParser<>("TermSuggestionEntryParser", true,
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/suggest/term/TermSuggestion.java]: Same

File: core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+import static org.elasticsearch.common.xcontent.XContentHelper.toXContent;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertToXContentEquivalent;
+
+public class SuggestionEntryTests extends ESTestCase {
+
+    @SuppressWarnings("rawtypes")
+    private static final Map<Class<? extends Entry>, Function<XContentParser, ? extends Entry>> ENTRY_PARSERS = new HashMap<>();
+    {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java]: Can you make this block static?

File: core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        ENTRY_PARSERS.put(Suggestion.Entry.class, Suggestion.Entry::fromXContent);
+        ENTRY_PARSERS.put(TermSuggestion.Entry.class, TermSuggestion.Entry::fromXContent);
+        ENTRY_PARSERS.put(PhraseSuggestion.Entry.class, PhraseSuggestion.Entry::fromXContent);
+        ENTRY_PARSERS.put(CompletionSuggestion.Entry.class, CompletionSuggestion.Entry::fromXContent);
+    }
+
+    /**
+     * Create a randomized Suggestion.Entry, determined by the type argument.
+     * @param type determines the entry type for this test. Valid arguments are 0 for {@link TermSuggestion.Entry},
+     * 1 for {@link PhraseSuggestion.Entry} and 2 for {@link CompletionSuggestion.Entry}
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java]: I think this is out of date?

File: core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        int offset = randomInt();
+        int length = randomInt();
+        @SuppressWarnings("rawtypes")
+        Entry entry = null;
+        Supplier<Option> supplier = null;
+        if (entryType == Suggestion.Entry.class) {
+            entry = new Suggestion.Entry<>(entryText, offset, length);
+            supplier = SuggestionOptionTests::createTestItem;
+        } else if (entryType == TermSuggestion.Entry.class) {
+                entry = new TermSuggestion.Entry(entryText, offset, length);
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java]: Wrong formating

File: core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        for (int i = 0; i < numOptions; i++) {
+            entry.addOption(supplier.get());
+        }
+        return entry;
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testFromXContent() throws IOException {
+        for (Class<? extends Entry> entryType : ENTRY_PARSERS.keySet()) {
+        Entry<Option> entry = createTestItem(entryType);
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/suggest/SuggestionEntryTests.java]: Wrong formating, or maybe you didn't want to iterate over each entry type and randomize it instead?

**Review by tlrx — CHANGES_REQUESTED** [review]
I left some comments but this is only cosmetics.

**Comment by cbuescher** [comment]
@tlrx thanks for the review, I addressed all of your comments, re-running also locally because CI seems stuck currently.

**Review by tlrx — APPROVED** [review]

**Comment by cbuescher** [comment]
Merged to 5.x with 7c229ce32bb93c6e6077626a81d838b250c2d54a

**Pull request merged**